test/e2e/framework/dataplane.go

```diff
@@ -125,3 +141,74 @@ func execCmdInBash(p tcp.ConnectivityTestParams, cmd []string, pod *v1.Pod) (str
 
 	return p.Framework.ExecWithOptions(execOptions, p.FromCluster)
 }
+
+func createTCPHeadlessService(cluster framework.ClusterIndex, selectorName string, port int, namespace string) *v1.Service {
+	tcpService := v1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: fmt.Sprintf("test-headless-svc-%s", selectorName),
+		},
+		Spec: v1.ServiceSpec{
+			Type:      "ClusterIP",
+			ClusterIP: v1.ClusterIPNone,
+			Ports: []v1.ServicePort{{
+				Port:       int32(port),
+				Name:       "tcp",
+				TargetPort: intstr.FromInt(port),
+				Protocol:   v1.ProtocolTCP,
+			}},
+			Selector: map[string]string{
+				testAppLabel: selectorName,
+			},
+		},
+	}
+
+	services := framework.KubeClients[cluster].CoreV1().Services(namespace)
+
+	return framework.AwaitUntil("create service", func() (interface{}, error) {
+		service, err := services.Create(context.TODO(), &tcpService, metav1.CreateOptions{})
+		if errors.IsAlreadyExists(err) {
+			err = services.Delete(context.TODO(), tcpService.Name, metav1.DeleteOptions{})
+			if err != nil {
+				return nil, err
+			}
+
+			service, err = services.Create(context.TODO(), &tcpService, metav1.CreateOptions{})
+		}
+
+		return service, err
+	}, framework.NoopCheckResult).(*v1.Service)
+}
+
+func getGlobalIngressIP(p tcp.ConnectivityTestParams, service *v1.Service) string {
+	if p.ToEndpointType == tcp.GlobalServiceIP {
+		return p.Framework.AwaitGlobalIngressIP(p.ToCluster, service.Name, service.Namespace)
+	} else if p.ToEndpointType == tcp.GlobalPodIP {
+		podList := awaitPodsByLabelSelector(p.ToCluster, labels.Set(service.Spec.Selector).AsSelector().String(),
+			service.Namespace, 1)
+		ingressIPName := fmt.Sprintf("pod-%s", podList.Items[0].Name)
+		return p.Framework.AwaitGlobalIngressIP(p.ToCluster, ingressIPName, service.Namespace)
+	}
+
+	return ""
+}
+
+func awaitPodsByLabelSelector(cluster framework.ClusterIndex, labelSelector, namespace string, expectedCount int) *v1.PodList {
```

Comment:
I assume this will be moved to shipyard?

Reply:
Yes. This and maybe HeadlessService too.

Reply:
I think we should just put it in shipyard now rather than adding technical debt - there's no rush in getting this PR in.